File: beecrowd/1023/hash_counting_sort.py
# Since n can go up to 10⁶, the brute force algorithm which has a complexity of 
# 2n + nlogn => 2*(10^6 + 10^7) => 20M, which for multiple test cases gives TLE.

# The bottleneck is any step that costs O(n*log n), in our case, sorting.
# Idea: We can avoid O(n*log n) sorting by using a hash table and counting sort
# because we know the avg consumption is always between [0, 200].

from collections import defaultdict

def counting_sort(table, high):
    present = [False] * (high + 1)
    sorted_arr = []

    for x in table:
        present[x] = True
    for i in range(high + 1):
        if present[i]:
            elem = table[i], i
            sorted_arr.append(elem)
    return sorted_arr

def main():
    case_num = 0
    while True:
        n = int(input())
        if n == 0:
            return
        case_num += 1
        if case_num != 1:
            print()

        table = defaultdict(int)
        total_consump = 0
        total_people = 0
        for _ in range(n):
            people, consump = map(int, input().split())
            avg_consump = consump // people
            table[avg_consump] += people
            total_consump += consump
            total_people += people
        sorted_arr = counting_sort(table, 200)
        avg = total_consump / total_people

        output = ' '.join([f'{people}-{consump}' for people, consump in sorted_arr])
        avg = int(avg * 100) / 100

        print(f'Cidade# {case_num}:')
        print(output)
        print(f'Consumo medio: {avg:.2f} m3.')

if __name__ == '__main__':
    main()

File: beecrowd/1023/hash_counting_sort.cpp
#include <bits/stdc++.h>
using namespace std;

struct DefaultDict {
  unordered_map<int, int> data;

  int& operator[](const int& key) {
    if (!data.contains(key))
      data[key] = 0;
    return data[key];
  }
};

vector<pair<int, int>> counting_sort(const DefaultDict& table, int high = 200) {
  vector<pair<int, int>> res;
  for (int i = 0; i <= high; i++) {
    auto p = table.data.find(i);
    if (p != table.data.end())
      res.push_back(*p);
  }
  return res;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int tc = 0, num_residencies, people, consump, avg_residence;
  float avg;

  cin >> num_residencies;
  while (num_residencies != 0) {
    DefaultDict num_residents_with_avg;
    int total_people = 0, total_consump = 0;

    tc++;
    if (tc != 1)
      cout << '\n';

    for (int i = 0; i < num_residencies; i++) {
      cin >> people >> consump;
      total_people += people;
      total_consump += consump;
      avg_residence = consump / people;
      num_residents_with_avg[avg_residence] += people;
    }
    avg = static_cast<float>(total_consump) / total_people;
    avg = int(avg * 100) / 100.0f;
    auto vals = counting_sort(num_residents_with_avg);

    cout << "Cidade# " << tc << ":\n";
    for (int i = 0; i < vals.size(); i++) {
      cout << vals[i].second << '-' << vals[i].first;
      if (i < vals.size() - 1)
        cout << ' ';
      else
        cout << '\n';
    }
    cout << "Consumo medio: " << fixed << setprecision(2) << avg << " m3.\n";

    cin >> num_residencies;
  }
}
